File: E-Commerce-Order-Processing-System-Perl/data/sqlite_explorer.pl
#!/usr/bin/env perl
use strict;
use warnings;
use DBI;
use Term::ReadLine;
#type sa terminal og perl sqlite_explorer.pl
#perl sqlite_explorer.pl path/to/database.db
sub explore_db {
    my $db_path = shift;
    
    # Connect to database
    my $dbh = DBI->connect("dbi:SQLite:dbname=$db_path", "", "", {
        RaiseError => 1,
        AutoCommit => 1,
    }) or die "Cannot connect to database: $DBI::errstr";
    
    print "\n" . "=" x 60 . "\n";
    print "SQLite Database: $db_path\n";
    print "=" x 60 . "\n\n";
    
    # List all tables
    my $tables = $dbh->selectall_arrayref(
        "SELECT name FROM sqlite_master WHERE type='table' ORDER BY name"
    );
    
    print "📊 Tables in database:\n";
    foreach my $table (@$tables) {
        my $table_name = $table->[0];
        my ($count) = $dbh->selectrow_array("SELECT COUNT(*) FROM $table_name");
        printf("  ✓ %-25s (%d rows)\n", $table_name, $count);
    }
    
    print "\n" . "=" x 60 . "\n";
    print "💡 Commands you can run:\n";
    print "  - To see table structure: PRAGMA table_info(table_name)\n";
    print "  - To query data: SELECT * FROM table_name LIMIT 10\n";
    print "  - Type 'exit' or 'quit' to exit\n";
    print "=" x 60 . "\n\n";
    
    # Interactive mode
    my $term = Term::ReadLine->new('SQLite Explorer');
    $term->ornaments(0);  # Disable fancy formatting
    
    while (defined(my $query = $term->readline('SQL> '))) {
        $query =~ s/^\s+|\s+$//g;  # Trim whitespace
        
        last if $query =~ /^(exit|quit|q|\.quit|\.exit)$/i;
        next if $query eq '';
        
        eval {
            my $sth = $dbh->prepare($query);
            $sth->execute();
            
            if ($query =~ /^\s*(SELECT|PRAGMA)/i) {
                my $names = $sth->{NAME};
                my @results;
                
                while (my $row = $sth->fetchrow_arrayref()) {
                    push @results, [@$row];
                }
                
                if (@results) {
                    # Print headers
                    if ($names && @$names) {
                        print "\n" . join(" | ", @$names) . "\n";
                        print "-" x (length(join(" | ", @$names)) + 10) . "\n";
                    }
                    
                    # Print rows
                    foreach my $row (@results) {
                        my @formatted = map { defined $_ ? $_ : 'NULL' } @$row;
                        print join(" | ", @formatted) . "\n";
                    }
                    print "\n(" . scalar(@results) . " rows)\n\n";
                } else {
                    print "(No results)\n\n";
                }
            } else {
                print "✓ Query executed successfully\n\n";
            }
            
            $sth->finish();
        };
        
        if ($@) {
            print "❌ Error: $@\n";
        }
    }
    
    print "\nGoodbye! 👋\n";
    $dbh->disconnect();
}

# Main
my $db_path = $ARGV[0] || "E-Commerce-Order-Processing-System-Perl/data/ecommerce.db";

if (!-f $db_path) {
    die "Error: Database file not found: $db_path\n";
}

explore_db($db_path);

File: E-Commerce-Order-Processing-System-Python/data/sqlite_explorer.py
import sqlite3
import sys
#type sa terminal og python sqlite_explorer.py

def explore_db(db_path):
    """Interactive SQLite database explorer"""
    try:
        conn = sqlite3.connect(db_path)
        cursor = conn.cursor()
        
        print(f"\n{'='*60}")
        print(f"SQLite Database: {db_path}")
        print(f"{'='*60}\n")
        
        # List all tables
        cursor.execute("SELECT name FROM sqlite_master WHERE type='table' ORDER BY name")
        tables = cursor.fetchall()
        
        print("📊 Tables in database:")
        for table in tables:
            table_name = table[0]
            cursor.execute(f"SELECT COUNT(*) FROM {table_name}")
            count = cursor.fetchone()[0]
            print(f"  ✓ {table_name:<25} ({count} rows)")
        
        print(f"\n{'='*60}")
        print("💡 Commands you can run:")
        print("  - To see table structure: PRAGMA table_info(table_name)")
        print("  - To query data: SELECT * FROM table_name LIMIT 10")
        print(f"{'='*60}\n")
        
        # Interactive mode
        while True:
            query = input("SQL> ").strip()
            
            if query.lower() in ['exit', 'quit', 'q', '.quit', '.exit']:
                print("Goodbye! 👋")
                break
            
            if not query:
                continue
                
            try:
                cursor.execute(query)
                
                if query.strip().upper().startswith('SELECT') or query.strip().upper().startswith('PRAGMA'):
                    results = cursor.fetchall()
                    
                    if results:
                        # Print column names
                        if cursor.description:
                            headers = [desc[0] for desc in cursor.description]
                            print("\n" + " | ".join(headers))
                            print("-" * (sum(len(h) for h in headers) + 3 * len(headers)))
                        
                        # Print rows
                        for row in results:
                            print(" | ".join(str(cell) for cell in row))
                        print(f"\n({len(results)} rows)\n")
                    else:
                        print("(No results)\n")
                else:
                    conn.commit()
                    print("✓ Query executed successfully\n")
                    
            except sqlite3.Error as e:
                print(f"❌ Error: {e}\n")
        
        conn.close()
        
    except Exception as e:
        print(f"❌ Error opening database: {e}")
        sys.exit(1)

if __name__ == "__main__":
    if len(sys.argv) > 1:
        db_path = sys.argv[1]
    else:
        # Default to Python ecommerce database
        db_path = "E-Commerce-Order-Processing-System-Python/data/ecommerce.db"
    
    explore_db(db_path)
